aten/src/ATen/native/cuda/MaxUnpooling.cu

```diff
@@ -0,0 +1,598 @@
+    PackedTensorAccessor<T, 4> gradInput,
+    int offsetZ) {
+  int iColumn = blockIdx.x * blockDim.x + threadIdx.x;
+  int iRow = blockIdx.y * blockDim.y + threadIdx.y;
+  int iFrame = (blockIdx.z + offsetZ) % gradInput.size(1); // output frame/time
+  int slice =
+      (blockIdx.z + offsetZ) / gradInput.size(1); // output slice/feature
+
+  if (iRow < gradInput.size(2) && iColumn < gradInput.size(3)) {
+    int64_t index = indices[slice][iFrame][iRow][iColumn];
+    T grad_val = gradOutputData[slice * oT * oH * oW + index];
+    gradInput[slice][iFrame][iRow][iColumn] = grad_val;
+  }
+}
+
+Tensor& max_unpooling2d_forward_out_cuda(
+    Tensor& output,
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size) {
+  AT_CHECK(
+      indices.scalar_type() == at::ScalarType::Long,
+      "elements in indices should be type Long");
+  auto owidth = output_size[0];
+  auto oheight = output_size[1];
+
+  TensorArg output_arg{output, "output", 1}, self_arg{self, "self", 2},
+      indices_arg{indices, "indices", 3};
+  checkAllSameGPU(
+      "max_unpooling2d_forward_out_cuda", {output_arg, self_arg, indices_arg});
+
+  AT_CHECK(self.numel() > 0, "Input must be non-empty tensor");
+
+  AT_CHECK(
+      (self.ndimension() == 3 || self.ndimension() == 4),
+      "Input to max_unpooling2d should be a 3d or 4d Tensor",
+      self.sizes());
+  AT_CHECK(
+      self.sizes() == indices.sizes(),
+      "Shape of input must match shape of indices");
+  AT_CHECK(
+      output_size.size() == 2,
+      "There should be exactly two elements (width, height) in output_size");
+
+  int64_t dimw = 2;
+  int64_t dimh = 1;
+  int64_t numBatch = 1;
+
+  int64_t numChannels;
+  int64_t inputHeight;
+  int64_t inputWidth;
+
+  if (self.ndimension() == 4) {
+    numBatch = self.size(0);
+    dimw++;
+    dimh++;
+  }
+  numChannels = self.size(dimh - 1);
+  inputHeight = self.size(dimh);
+  inputWidth = self.size(dimw);
+
+  auto input_contiguous = self.contiguous();
+  auto indices_contiguous = indices.contiguous();
+
+  output.resize_({numBatch, numChannels, oheight, owidth});
+
+  output.zero_();
+
+  dim3 block(512);
+  dim3 grid((output.numel() + 512 - 1) / 512);
+
+  AT_DISPATCH_ALL_TYPES_AND_HALF(
+      self.type(), "max_unpooling2d_forward_kernel", ([&] {
+        max_unpooling2d_forward_kernel<<<
+            grid,
+            block,
+            0,
+            at::cuda::getCurrentCUDAStream()>>>(
+            self.numel(),
+            input_contiguous.data<scalar_t>(),
+            indices_contiguous.data<int64_t>(),
+            numChannels,
+            inputHeight,
+            inputWidth,
+            oheight,
+            owidth,
+            output.data<scalar_t>());
+      }));
+  AT_CHECK(
+      cudaGetLastError() == cudaSuccess,
+      "RoiPooling2d_forward_kernel failed with error code ",
+      cudaGetLastError());
+  if (self.ndimension() == 3) {
+    output.resize_({numChannels, oheight, owidth});
```

Comment:
This is not your fault (because the old code did it this way too), but restriding the tensor like this makes my skin crawl. uuuuuuu (The reason I hate it is because no resize should ever actually happen; the numels match and so this is a re-viewing operation.)